cassovary-core/src/main/scala/com/twitter/cassovary/graph/MemoryMappedDirectedGraph.scala

```diff
@@ -0,0 +1,153 @@
+
+/**
+ * A graph which reads edge data from a memory mapped file.  There is no object overhead per node: the memory
+ * used for n nodes and m edges with both in-neighbor and out-neighbor access is exactly 8 + 16*n + 8*m bytes.
+ * Also, Loading is very fast because no parsing of text is required.  Loading time is exactly the time it takes the operating system
+ * to map data from disk into memory.
+ *
+ * Nodes are numbered sequentially from 0 to nodeCount - 1 and must be a range of this form (i.e. nodeCount == maxNodeId + 1).
+ * When transforming a graph where nodeCount <= maxNodeId to this format, new nodes with no neighbors will be implicitly created.
+ *
+ * Currently only supports storing both in-neighbors and out-neighbors of nodes (StoredGraphDir.BothInOut).
+ * The binary format is currently subject to change.
+ */
+
+/* Storage format
+byteCount  data
+4          (reserved, later use for versioning or indicating undirected vs directed)
+4          n (i. e. the number of nodes)
+8*(n+1)    Offsets into out-neighbor data. Index i (a Long) points to the out-neighbor data of node i.
+           Index n is needed to compute the outdegree of node n - 1.
+8*(n+1)    Offsets into in-neighbor data (Longs) (Same interpretation as out-neighbor offsets)
+m          out-neighbor data
+m          in-neighbor data
+ */
+class MemoryMappedDirectedGraph(file: File) extends DirectedGraph[Node] {
+  val data: IntLongSource = new MemoryMappedIntLongSource(file)
+
+  override val nodeCount = data.getInt(4)
+
+  private def outboundOffset(id: Int): Long = data.getLong(8L + 8L * id)
+
+  private def outDegree(id: Int): Int = ((outboundOffset(id + 1) - outboundOffset(id)) / 4).toInt
+
+  private def inboundOffset(id: Int): Long = data.getLong(8L + 8L * (nodeCount + 1) + 8L * id)
+
+  private def inDegree(id: Int): Int = ((inboundOffset(id + 1) - inboundOffset(id)) / 4).toInt
+
+  /* Only created when needed (there is no array of these stored). */
+  private class MemoryMappedDirectedNode(override val id: Int) extends Node {
+    val nodeOutboundOffset = outboundOffset(id)
+    val nodeInboundOffset = inboundOffset(id)
+    override def outboundNodes(): Seq[Int] = new IndexedSeq[Int] {
+      val length: Int = outDegree(id)
+      def apply(i: Int): Int =  data.getInt(nodeOutboundOffset + 4L * i)
+    }
+    override def inboundNodes(): Seq[Int] = new IndexedSeq[Int] {
+      val length: Int = inDegree(id)
+      def apply(i: Int): Int =  data.getInt(nodeInboundOffset + 4L * i)
+    }
+  }
+
+  override def getNodeById(id: Int): Option[Node] =
+    if(0 <= id && id < nodeCount) {
+      Some(new MemoryMappedDirectedNode(id))
+    } else {
+      None
+    }
+
+  override def iterator: Iterator[Node] = (0 to nodeCount).iterator flatMap (i => getNodeById(i))
+
+  override def edgeCount: Long = outboundOffset(nodeCount) - outboundOffset(0)
+
+  // Uncomment in future if maxNodeId becomes a method:
```

Comment:
I don't want to forget about this, though, because right now I think calling maxNodeId will silently create millions of Node objects.  Is it OK if I leave this, and make a PR to make maxNodeId a method soon? I also need maxNodeId to be a method for the dynamic graphs. (At Teapot we've seen confusing behavior from maxNodeId on dynamic graphs).